File: apps/quickink/android/app/src/main/java/app/quickink/mobile/data/capturelocation/CaptureLocationEntity.kt
/*
 * CaptureLocationEntity.kt
 *
 * Room @Entity for the `capture_locations` many-to-many join — links
 * captures to user-defined locations. Each row syncs independently
 * (own id + dirty + tombstone trio) so a location added on phone A
 * reaches phone B without re-syncing the entire capture. Mirror of
 * CaptureTagEntity.
 */

package app.quickink.mobile.data.capturelocation

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import app.quickink.mobile.data.capture.CaptureEntity
import app.quickink.mobile.data.location.LocationEntity

@Entity(
    tableName = "capture_locations",
    foreignKeys = [
        ForeignKey(
            entity = CaptureEntity::class,
            parentColumns = ["id"],
            childColumns = ["capture_id"],
            onDelete = ForeignKey.NO_ACTION,
        ),
        ForeignKey(
            entity = LocationEntity::class,
            parentColumns = ["id"],
            childColumns = ["location_id"],
            onDelete = ForeignKey.NO_ACTION,
        ),
    ],
    indices = [
        Index(value = ["capture_id"], name = "idx_capture_locations_capture"),
        Index(value = ["location_id"], name = "idx_capture_locations_location"),
        Index(value = ["dirty"], name = "idx_capture_locations_dirty"),
        Index(value = ["deleted_at"], name = "idx_capture_locations_tombstone"),
        Index(value = ["capture_id", "location_id"], name = "idx_capture_locations_pair"),
    ],
)
data class CaptureLocationEntity(
    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String,

    @ColumnInfo(name = "capture_id")
    val captureId: String,

    @ColumnInfo(name = "location_id")
    val locationId: String,

    /**
     * Provenance of this location assignment. `"manual"` is the
     * common case (user picked it from the location chip rail);
     * future sources land without a migration.
     */
    @ColumnInfo(name = "source", defaultValue = "manual")
    val source: String = "manual",

    @ColumnInfo(name = "drive_file_id")
    val driveFileId: String? = null,

    @ColumnInfo(name = "created_at")
    val createdAt: String,

    @ColumnInfo(name = "updated_at")
    val updatedAt: String,

    @ColumnInfo(name = "dirty", defaultValue = "1")
    val dirty: Boolean = true,

    @ColumnInfo(name = "deleted_at")
    val deletedAt: String? = null,
)

File: apps/quickink/ios/QuickInk/Data/CaptureLocationEntity.swift
/*
 * CaptureLocationEntity.swift
 *
 * GRDB record for the `capture_locations` many-to-many join — links
 * a capture to a user-defined location. Each row syncs independently
 * (own id + dirty + tombstone trio) so a location attached on phone
 * A reaches phone B without re-syncing the parent capture. Mirror of
 * `CaptureLocationEntity.kt`.
 */

import Foundation
import GRDB

public struct CaptureLocationEntity: Codable, FetchableRecord, PersistableRecord, Equatable, Sendable, Identifiable {

    public static let databaseTableName = "capture_locations"

    public var id: String
    public var captureId: String
    public var locationId: String
    public var source: String
    public var driveFileId: String?
    public var createdAt: String
    public var updatedAt: String
    public var dirty: Bool
    public var deletedAt: String?

    public init(
        id: String,
        captureId: String,
        locationId: String,
        source: String = "manual",
        driveFileId: String? = nil,
        createdAt: String,
        updatedAt: String,
        dirty: Bool = true,
        deletedAt: String? = nil
    ) {
        self.id          = id
        self.captureId   = captureId
        self.locationId  = locationId
        self.source      = source
        self.driveFileId = driveFileId
        self.createdAt   = createdAt
        self.updatedAt   = updatedAt
        self.dirty       = dirty
        self.deletedAt   = deletedAt
    }

    public enum CodingKeys: String, CodingKey {
        case id
        case captureId   = "capture_id"
        case locationId  = "location_id"
        case source
        case driveFileId = "drive_file_id"
        case createdAt   = "created_at"
        case updatedAt   = "updated_at"
        case dirty
        case deletedAt   = "deleted_at"
    }
}
